samples/webcodecs-echo/js/main.js

```diff
@@ -0,0 +1,294 @@
+  inputStream.cancel();
+  outputStream.abort(); 
+  addToEventLog('stop(): input stream cancelled and output stream aborted');
+}
+
+document.addEventListener('DOMContentLoaded', async function(event) {
+  if (stopped) return;
+  addToEventLog('DOM Content Loaded');
+
+  if (typeof MediaStreamTrackProcessor === 'undefined' ||
+      typeof MediaStreamTrackGenerator === 'undefined') {
+    addToEventLog('Your browser does not support the experimental Mediacapture-transform API.\n' +
+        'Please launch with the --enable-blink-features=WebCodecs,MediaStreamInsertableStreams flag','fatal');
+    return;
+  }
+
+  if (typeof WebTransport === 'undefined') {
+    addToEventLog('Your browser does not support the WebTransport API.', 'fatal');
+    return;
+  }
+
+  try {
+    gotDevices(await navigator.mediaDevices.enumerateDevices());
+  } catch (e) {
+    addToEventLog('Error in Device enumeration');
+  }
+  constraints.deviceId = videoSource ? {exact: videoSource} : undefined;
+  //addToEventLog('videoSource: ' + JSON.stringify(videoSource) + ' DeviceId: ' + constraints.deviceId);
+  // Get a MediaStream from the webcam.
+  mediaStream = await navigator.mediaDevices.getUserMedia(constraints);
+  // Connect the webcam stream to the video element.
+  document.getElementById('inputVideo').srcObject = mediaStream;
+  // Create a new worker.
+  streamWorker = new Worker("js/stream_worker.js");
+  addToEventLog('Worker created.');
+  // Print messages from the worker in the text area.
+  streamWorker.addEventListener('message', function(e) {
+    addToEventLog('Worker msg: ' + e.data.text, e.data.severity);
+  }, false);
+
+  stopButton.onclick = () => {
+    addToEventLog('Stop button clicked.');
+    stop();
+  }
+
+  connectButton.onclick = () => {
+    connectButton.disabled = true;
+    stopButton.disabled = false;
+    hwButtons.style.display = "none";
+    prefButtons.style.display = "none";
+    codecButtons.style.display = "none";
+    resButtons.style.display = "none";
+    modeButtons.style.display = "none";
+    rateInput.style.display = "none";
+    keyInput.style.display = "none";
+    startMedia();
+  }
+
+  async function startMedia() {
+    if (stopped) return;
+    addToEventLog('startMedia called'); 
+    try {
+      // Collect the bitrate
+      const rate = document.getElementById('rate').value;
+
+      // Collect the keyframe gap
+      const keygap = document.getElementById('keygap').value;
+
+      // Create a MediaStreamTrackProcessor, which exposes frames from the track
+      // as a ReadableStream of VideoFrames.
+      let [track] = mediaStream.getVideoTracks();
+      let ts = track.getSettings();
+      const processor = new MediaStreamTrackProcessor(track);
+      inputStream = processor.readable;
+
+      // Create a MediaStreamTrackGenerator, which exposes a track from a
+      // WritableStream of VideoFrames.
+      const generator = new MediaStreamTrackGenerator({kind: 'video'});
```

Comment:
Please add a comment that this is using non-standard Chrome-only versions of *MediaStreamTrackProcessor* and *VideoTrackGenerator* for now, and creating them on main-thread not on the worker, due to lack of transferable MediaStreamTrack in any implementation.

Maybe someday I'll attempt a polyfill in adapter.js...

Reply:
Fixed.